File: leetcode/divide_two_integers.rb
# https://leetcode.com/problems/divide-two-integers/
#
# @param {Integer} dividend
# @param {Integer} divisor
# @return {Integer}
def divide(dividend, divisor)
    return 0 if dividend == 0

    is_neg = true if (dividend < 0 && divisor > 0) || (divisor < 0 && dividend > 0)

    dividend = dividend.abs
    divisor = divisor.abs

    res = 0

    while dividend >= divisor

        val = divisor
        mul = 1

        while true

            new_val = val + val

            if dividend >= new_val
                val = new_val
                mul += mul
            else
                break
            end
        end

        res += mul
        dividend -= val

    end

    res = is_neg ? -res : res

    max = (2**31) - 1
    res > max ? max : res

end

puts divide(10, 3) # 3

puts divide(7, -3) # -2

puts divide(-10, -3) # 3

File: leetcode/DivideTwoIntegers.java
/*
	https://leetcode.com/problems/divide-two-integers/
*/
class DivideTwoIntegers {
	public static void main(String[] args) {
		System.out.println(new Solution().divide(10, 3)); // 3

		System.out.println(new Solution().divide(2147483647, 1)); // 2147483647
		System.out.println(new Solution().divide(-2147483648, 1)); // -2147483648
		System.out.println(new Solution().divide(-2147483648, -1)); // 2147483647
	}
}

class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1){
        	return Integer.MAX_VALUE;
        }

        boolean isNegative = false;
        if ((dividend < 0 && divisor > 0) || (divisor < 0 && dividend > 0)){
        	isNegative = true;
        }

        long ldividend = Math.abs((long)dividend);
        long ldivisor = Math.abs((long)divisor);

        long result = divide(ldividend, ldivisor);

        return (int)(isNegative ? -result : result);
    }

    private long divide(long dividend, long divisor) {
        if (divisor == 1) return dividend;
        if (dividend < divisor) return 0;

        long curr = divisor;
        long result = 1;

        while(curr + curr <= dividend ) {
            curr += curr;
            result += result;
        }

        return result + divide(dividend - curr, divisor);
    }
}
